src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.cpp

```diff
@@ -0,0 +1,910 @@
+        apply_rec(loop_id);
+      }
+    }
+
+    // Compute set of locals at function level
+    top_level_local.insert(
+      goto_function.parameter_identifiers.begin(),
+      goto_function.parameter_identifiers.end());
+
+    for(const auto &entry : ident_to_loop_id_map)
+    {
+      if(entry.second == loop_info_map.size())
+      {
+        top_level_local.insert(entry.first);
+      }
+    }
+
+    // Compute subset of function locals that must be tracked in the function's
+    // write set. A local must be tracked if it is dirty or if it may be
+    // assigned by one of the top level loops.
+    for(const auto &ident : top_level_local)
+    {
+      if(dirty(ident))
+      {
+        top_level_tracked.insert(ident);
+      }
+      else
+      {
+        // Check if this ident is touched by one of the top level loops
+        for(const auto loop_id : top_level_loops)
+        {
+          if(is_assigned(ident, loop_info_map.at(loop_id).assigns))
+          {
+            top_level_tracked.insert(ident);
+          }
+        }
+      }
+    }
+  }
+
+  void apply_rec(const std::size_t loop_id)
+  {
+    // update the DECL ident -> loop id map on the way down
+    for(auto target : loop_nesting_graph[loop_id].instructions)
+    {
+      if(target->is_decl())
+      {
+        ident_to_loop_id_map[target->decl_symbol().get_identifier()] = loop_id;
+      }
+    }
+
+    // Process inner loops
+    for(const auto inner_loop_id : loop_nesting_graph.get_predecessors(loop_id))
+    {
+      apply_rec(inner_loop_id);
+    }
+
+    auto &loop = loop_nesting_graph[loop_id];
+
+    // Collect identifiers that are local to this loop only
+    // (excluding nested loop).
+    std::unordered_set<irep_idt> loop_local;
+    for(const auto &target : loop.instructions)
+    {
+      if(target->is_decl())
+      {
+        auto &ident = target->decl_symbol().get_identifier();
+        if(ident_to_loop_id_map[ident] == loop_id)
+          loop_local.insert(ident);
+      }
+    }
+
+    // Compute subset of locals that must be tracked in the loop's write set.
+    // A local must be tracked if it is dirty or if it may be assigned by one
+    // of the inner loops.
+    std::unordered_set<irep_idt> loop_tracked;
+    for(const auto &ident : loop_local)
+    {
+      if(dirty(ident))
+      {
+        loop_tracked.insert(ident);
+      }
+      else
+      {
+        // Check if this ident is touched by one of the inner loops
+        for(const auto inner_loop_id :
+            loop_nesting_graph.get_predecessors(loop_id))
+        {
+          if(is_assigned(ident, loop_info_map.at(inner_loop_id).assigns))
+            loop_tracked.insert(ident);
+        }
+      }
+    }
+
+    // Process loop contract clauses
+    exprt::operandst invariant_clauses = get_invariants(loop.latch);
+    exprt::operandst decreases_clauses = get_decreases(loop.latch);
+    exprt::operandst assigns_clauses = get_assigns(loop.latch);
+    std::set<exprt> assigns;
```

Comment:
If it is important that this is the same type as `assignst` then it should be `assignst` here instead of `std::set<exprt>`. Else it should really be `std::unordered_set<exprt, irep_hash>`.